File: leetcode/5-longest-palindromic-substring.js
// https://leetcode.com/problems/longest-palindromic-substring/ (top-interview-questions)

/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
  
  let palindrome = "";

  for (let i = 0; i < s.length; i++) {
    const longestPalAroundI = longestPalindromAround(i, s);
    if ( palindrome.length < longestPalAroundI.length ) {
      palindrome = longestPalAroundI;
    }
  }

  return palindrome;
};


const longestPalindromAround = function (i, s) {

  let palindrome = "";
  
  // check odd length palindroms
  let l = i;
  let r = i;
  while ( l >= 0 && r < s.length && s[l] === s[r]) {
    if (palindrome.length < (r-l+1)) {
      palindrome = s.slice(l, r+1);
    }
    l -= 1;
    r += 1;
  }

  // check for even length palindroms
  l = i;
  r = i + 1;
  while (l >= 0 && r < s.length && s[l] === s[r]) {
    if (palindrome.length < (r-l+1)) {
      palindrome = s.slice(l, r+1);
    }
    l -= 1;
    r += 1;
  }

  return palindrome;
}

File: leetcode/5-longest-palindromic-substring.py
# https://leetcode.com/problems/longest-palindromic-substring/ (top-interview-questions)

class Solution:
    def longestPalindrome(self, s: str) -> str:
        palindrome = ''
        length = 0

        for i in range(len(s)):
            # check odd palindroms
            l, r = i, i
            # two-pointer approach to expand around each character in the string
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > length:
                    palindrome = s[l:r+1]
                    length = r-l+1
                l -= 1
                r += 1
            
            # check even palindroms
            l, r = i, i+1
            # two-pointer approach to expand around each character in the string
            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r - l + 1) > length:
                    palindrome = s[l:r+1]
                    length = r - l + 1
                l -= 1
                r += 1

        return palindrome